### DISCUSSION TITLE
Editor stable-like hit-animations off

### DISCUSSION THREAD

**Discussion by maotovisk** [ppy/osu#14298] [Q&A]
## Introduction
Hello, I've been working on a prototype (at https://github.com/maotovisk/osu/tree/stable-editor-hitanimations) to make the editor hit-animations (disabled hit animations) more like the stable editor for a while, since the current implementation lacks visual clue of the object being hit (even with the approach circle expanding) and the fade-out doesn't look like the current fade-out duration (and easing) we have on stable. 

### Stable implementation
https://user-images.githubusercontent.com/20617307/129452802-62a9fc48-c466-4217-8683-ea2acfd5f39c.mp4

### Lazer implementation (2021.809.0-lazer)

https://user-images.githubusercontent.com/20617307/129452775-1e2d9de3-6f6b-485a-bd25-7580b55bb74d.mp4

With the current implementation we have, there are some things that could be considered unwilled such as:

- The fade-out durations and orders (sliderhead, body and tail/repeats) misrepresent the mapper intentions regarding visual overlaps (notes or sliders that overlaps each other at the playfield).

- The kiai flashing also misrepresent overlaps because of the opacity change between the beats (mappers don't map with the music playing, they manually move the timeline to the next beat using scroll).
 
- The lack of visual clue of hitting the notes, such as the accent colour change on hit.

- (unrelated to hit-animations and already tracked? at https://github.com/ppy/osu/issues/10909) The autoplay clock of the editor is reduced causing inaccurate animations to be reproduced.

### The WIP Prototype

https://user-images.githubusercontent.com/20617307/129453031-a2809da3-8982-4ca2-bd03-c38cc314647a.mp4

## The issues with the prototype

This implementation fixes two of the main differences between the implementations we have on stable and lazer, but also introduces an issue that i'm not able to figure out how to fix: When changing skins (or enabling map combo colours) the already-hit circlepieces (both sliders' and hitcircles') go back to the accent colour they had when they got hit.

https://user-images.githubusercontent.com/20617307/129455126-cbc6c46d-96b5-4b4b-827c-bd7507dca62e.mp4

> ignore this flickering effect with the approach circles when playing the session for the first time, it's an issue introduced with recent changes(https://github.com/ppy/osu/issues/14315) and unrelated to the prototype.

It happens because I used the `hitobject.TransformBindableTo()`  method to bind the hitobject `AccentColour` to `White`  at the `StateUpdateTime` i'm not sure how to fix it, since this is the only way to change the `AccentColour`.

And I also don't have idea of how to disable the kiai flash animations and it surely has to have an alternative of its visual indicator if disabled.

I would appreciate some help and thoughts over this and sorry for taking your time.

**Comment by frenzibyte** [comment]
> And I also don't have idea of how to disable the kiai flash animations and it surely has to have an alternative of its visual indicator if disabled.

One way would be to expose the kiai flash drawable to `CirclePiece` (via an interface, `IMainCirclePiece`) that would point to the `KiaiFlash` drawable in case of `MainCirclePiece` and `KiaiFlashingSprite.flashingSprite` in case of `LegacyMainCirclePiece`, and hide it from there (via `((IMainCirclePiece)mainPieceContainer.CirclePiece.Drawable).KiaiFlash.Hide()`), inside the `ScheduleAfterChildren` of course.

A bit WTF of a way to do it, but can't think of anything else. And the above way would match with what `DrawableOsuEditorRuleset` intends to do. (performing logic all locally, rather than writing code in the implementations themselves)
